stew/io2.nim

```diff
@@ -1312,3 +1361,112 @@ proc readAllChars*(pathName: string): IoResult[string] =
+  when defined(posix):
+    let ltype =
+      case kind
+      of LockType.Shared:
+        cshort(posix.F_RDLCK)
+      of LockType.Exclusive:
+        cshort(posix.F_WRLCK)
+    var flockObj = FlockStruct(ltype: ltype, lwhence: cshort(posix.SEEK_SET),
+                               start: offset, length: size)
+    while true:
+      let res = posix.fcntl(cint(handle), posix.F_SETLK, addr flockObj)
+      if res == -1:
+        let errCode = ioLastError()
+        if errCode == EINTR:
+          continue
+        else:
+          return err(errCode)
+      else:
+        return ok()
+  elif defined(windows):
+    let (lowOffsetPart, highOffsetPart) = makeUint32(uint64(offset))
+    let (lowSizePart, highSizePart) = makeUint32(uint64(size))
+    var ovl = OVERLAPPED(offset: lowOffsetPart, offsetHigh: highOffsetPart)
+    let
+      flags =
+        case kind
+        of LockType.Shared:
+          LOCKFILE_FAIL_IMMEDIATELY
+        of LockType.Exclusive:
+          LOCKFILE_FAIL_IMMEDIATELY or LOCKFILE_EXCLUSIVE_LOCK
+      res = lockFileEx(uint(handle), flags, 0'u32, lowSizePart,
+                       highSizePart, addr ovl)
+    if res == 0:
+      err(ioLastError())
+    else:
+      ok()
+
+proc unlockFile*(handle: IoHandle, offset: int, size: int): IoResult[void] =
+  ## Clear shared or exclusive file segment lock for file handle ``handle`` and
+  ## range specified by ``offset`` and ``size`` parameters.
+  when defined(posix):
+    let ltype = cshort(posix.F_UNLCK)
+    var flockObj = FlockStruct(ltype: ltype, lwhence: cshort(posix.SEEK_SET),
+                               start: offset, length: size)
+    while true:
+      let res = posix.fcntl(cint(handle), F_SETLK, addr flockObj)
+      if res == -1:
+        let errCode = ioLastError()
+        if errCode == EINTR:
+          continue
+        else:
+          return err(errCode)
+      else:
+        return ok()
+  elif defined(windows):
+    let
+      (lowOffsetPart, highOffsetPart) = makeUint32(uint64(offset))
+      (lowSizePart, highSizePart) = makeUint32(uint64(size))
+    var ovl = OVERLAPPED(offset: lowOffsetPart, offsetHigh: highOffsetPart)
+    let res = unlockFileEx(uint(handle), 0'u32, lowSizePart,
+                           highSizePart, addr ovl)
+    if res == 0:
+      err(ioLastError())
+    else:
+      ok()
+
+proc lockFile*(handle: IoHandle, kind: LockType): IoResult[IoLockHandle] =
+  ## Apply exclusive or shared lock to whole file specified by file handle
+  ## ``handle``.
+  ##
+  ## ``kind`` - type of lock (shared or exclusive). Please not that only
+  ## exclusive locks has cross-platform compatible behavior. But exclusive
+  ## locks require ``handle`` to be opened for writing.
+  ##
+  ## On success returns ``IoLockHandle`` object which could be used for unlock.
+  let size =
+    block:
+      let res = ? getFileSize(handle)
+      when sizeof(int) == 4:
+        if res > int64(high(int)):
+          high(int)
```

Comment:
If `lockFile` is updated to use `Off` or `uint64` as arg, this could be replaced with an assert, as `getFileSize` uses `fstat` which returns an `off_t`, and `lockFile` uses `fcntl` which takes an `off_t`.

Reply:
This code protects from Windows overflow, because on Windows filesize is always 64bit even if system is only 32bit.

Reply:
But why would this overflow on Windows? Since `lockFileEx` also takes 64-bit inputs through the "high word", "low word" mechanism, we can just use `int64` for the offset value and this will be compiled to the correct code even on 32-bit systems, no? (since the C compilers support emulated types such as `__int64`)

Reply:
`io2.lockFile()` accepts `size` as `int` which could be `int32` and `int64` depends on system architecture. So if we running on Windows 32bit we can't use `getFileSize(handle)` result value which is always `int64` because there will be overflow when i'm going to convert `int32(int64 value)`. 
In Linux/BSD field of `stat` structure called `st_size` has `off_t` type which depends on architecture so it will be equal to `int32` in 32bit system. It means that Posix systems do not support files sizes bigger than high(int32). So in this case we going to convert only `int32(int32)` - so its safe.

Reply:
Shouldn't `io2.lockFile()` take the same type that `getFileSize(handle)` result returns? Otherwise, locking may silently fall back to a partial file lock for large files.

Is there a situation where `getFileSize` can return a value that's unsupported for locking?